internal/app/cache/cache.go

```diff
@@ -26,6 +26,14 @@ type Cache interface {
 
 	// DeleteRequest removes the given request from any cache entries it's present in.
 	DeleteRequest(key string, req *v2.DiscoveryRequest) error
+
+	// GetReadOnlyCache returns a copy of the cache that only exposes read-only methods in its interface.
+	GetReadOnlyCache() ReadOnlyCache
+}
+
+type ReadOnlyCache interface {
+	// Fetch returns the cached resource if it exists.
+	Fetch(key string) (*Resource, error)
```

Comment:
Should this return `Resource` rather than `*Resource` so we don't accidentally mutate the returned values?